src/pytest_memray/marks.py

```diff
@@ -51,28 +54,93 @@ def section(self) -> PytestSection:
-        total_memory_str = sizeof_fmt(self.total_allocated_memory)
-        max_memory_str = sizeof_fmt(self.max_memory)
-        return f"Test was limited to {max_memory_str} but allocated {total_memory_str}"
+        return f"Test was limited to {sizeof_fmt(self.max_memory)} but allocated {sizeof_fmt(self.total_allocated_memory)}"
+
+
+@dataclass
+class _LeakedInfo(_MemoryInfo):
+    """Type that holds leaked memory-related info for a failed test."""
+
+    @property
+    def section(self) -> PytestSection:
+        """Return a tuple in the format expected by section reporters."""
+        return (
+            "memray-leaked-memory",
+            self._generate_section_text("Test leaked", "List of leaked allocations:"),
+        )
+
+    @property
+    def long_repr(self) -> str:
+        """Generate a longrepr user-facing error message."""
+        return (
+            f"Test was allowed to leak {sizeof_fmt(self.max_memory)} "
+            "per location but at least one location leaked more"
+        )
 
 
 def limit_memory(
-    limit: str, *, _allocations: list[AllocationRecord], _config: Config
+    limit: str, *, _result_file: Path, _config: Config
 ) -> _MemoryInfo | None:
     """Limit memory used by the test."""
+    reader = FileReader(_result_file)
+    func = reader.get_high_watermark_allocation_records
+    allocations: list[AllocationRecord] = list((func(merge_threads=True)))
     max_memory = parse_memory_string(limit)
-    total_allocated_memory = sum(record.size for record in _allocations)
+    total_allocated_memory = sum(record.size for record in allocations)
     if total_allocated_memory < max_memory:
         return None
     num_stacks: int = cast(int, value_or_ini(_config, "stacks"))
     native_stacks: bool = cast(bool, value_or_ini(_config, "native"))
     return _MemoryInfo(
-        max_memory, total_allocated_memory, _allocations, num_stacks, native_stacks
+        max_memory, total_allocated_memory, allocations, num_stacks, native_stacks
+    )
+
+
+def limit_leaks(
+    location_limit: str,
+    *,
+    filter_fn: Optional[LeaksFilteringFunction] = None,
+    _result_file: Path,
+    _config: Config,
+) -> _MemoryInfo | None:
+    reader = FileReader(_result_file)
+    func = reader.get_leaked_allocation_records
+    allocations: list[AllocationRecord] = list((func(merge_threads=True)))
+
+    memory_limit = parse_memory_string(location_limit)
+
+    leaked_allocations = list(
+        allocation
+        for allocation in allocations
+        if (
+            allocation.size >= memory_limit
+            and (filter_fn is None or filter_fn(allocation.hybrid_stack_trace()))
+        )
+    )
+    if not leaked_allocations:
+        return None
+    total_leaked_memory = sum(allocation.size for allocation in leaked_allocations)
+
+    num_stacks: int = cast(int, value_or_ini(_config, "stacks"))
+    return _LeakedInfo(
+        memory_limit,
+        total_leaked_memory,
```

Comment:
This assigns `total_leaked_memory` to the `total_allocated_memory` attribute of the `_LeakedInfo`. That's not terribly clean, and confused me a bit while I was reading this code.

Reply:
I have refactored the common parts into a base dataclass and created 2 different downstream class in the inheritance. Tell me if you find some alternative approach cleaner or more suitable for this